extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -52,9 +60,12 @@ final class S3ChannelContext extends BaseSeekableChannelContext implements Seeka
     private S3Uri uri;
 
     /**
-     * Used to cache recently fetched fragments from the {@link #uri} for faster lookup.
+     * Used to cache recently fetched fragments from the {@link #uri} for faster lookup. Note that this cache can be
+     * shared across multiple contexts and be accessed concurrently.
      */
-    private final Request[] requests;
+    private final S3RequestCache sharedCache;
+
+    private final Map<Integer, Request> localCache;
```

Comment:
This could still be an array?

Reply:
I moved from array to map because every time a context object gets closed, we have to iterate over all existing cached requests and release them. 
So for an array, it will take extra iterations if not many requests are cached. That is why I kept a map.